src/core/highlight.ts

```diff
@@ -0,0 +1,55 @@
+import { Token } from './token';
+import singleton, { type Prism } from './prism';
+import { stringify } from './stringify';
+import type { Grammar, GrammarToken, GrammarTokens, RegExpLike } from '../types';
+
+/**
+ * Low-level function, only use if you know what you’re doing. It accepts a string of text as input
+ * and the language definitions to use, and returns a string with the HTML produced.
+ *
+ * The following hooks will be run:
+ * 1. `before-tokenize`
+ * 2. `after-tokenize`
+ * 3. `wrap`: On each {@link Token}.
+ *
+ * @param text A string with the code to be highlighted.
+ * @param language The name of the language definition passed to `grammar`.
+ * @param options An object containing the tokens to use.
+ *
+ * Usually a language definition like `Prism.languages.markup`.
+ * @returns The highlighted HTML.
+ * @example
+ * Prism.highlight('var foo = true;', 'javascript');
+ */
+export function highlight (
+	this: Prism,
+	text: string,
+	language: string,
+	options: HighlightOptions = {}
+): string {
+	const prism = this ?? singleton;
```

Comment:
We introduce a cycle by importing `singleton` in all' highlight' methods. This is how it looks:

```
src/core/prism.ts
  └──> src/core/classes/prism.ts
         └──> src/core/highlight.ts
                └──> src/core/prism.ts
```

Moving it to the file with the Prism class won't help. We probably need to break it somehow. Possibly, this is not only the issue in the build step. It might cause some other issues with types? 🤔

Reply:
Cycles are not a problem in ESM, they happen all the time. Does Rollup still struggle with them? I seem to recall it had an issue in the past.

Reply:
It logs a warning, but builds the code fine.

Reply:
Cool, no issue then? :D